libtransmission/peer-mgr.cc

```diff
@@ -396,18 +472,89 @@ class tr_swarm
             /* noop */
             break;
 
-        case tr_peer_event::Type::ClientGotRej:
-            s->active_requests.remove(s->tor->piece_loc(event.pieceIndex, event.offset).block, peer);
-            break;
-
         case tr_peer_event::Type::ClientGotChoke:
-            s->active_requests.remove(peer);
+            s->active_requests.remove(msgs);
             break;
 
         case tr_peer_event::Type::ClientGotPort:
-            if (auto* const info = peer->peer_info; info != nullptr)
+            // If we don't know the listening port of this peer (i.e. incoming connection and first time ClientGotPort)
+            if (auto* const info = msgs->peer_info; std::empty(msgs->peer_info->listen_port()))
             {
-                info->port() = event.port;
+                TR_ASSERT(info->is_connected());
+
+                auto nh = s->incoming_peer_info.extract(msgs->socket_address());
+                TR_ASSERT(!std::empty(nh));
+
+                // the info pointer is invalid and shouldn't be used from this point on:
+                // https://en.cppreference.com/w/cpp/container/unordered_map/extract
+                // Pointers and references to the extracted element remain valid, but cannot be used while element
+                // is owned by a node handle: they become usable if the element is inserted into a container.
+
+                auto& info_new = nh.mapped();
+                TR_ASSERT(info_new.is_connected());
+                TR_ASSERT(nh.key().address() == info_new.listen_address());
+                TR_ASSERT(std::empty(info_new.listen_port()));
+
+                // If we already know about this peer, merge the info objects without invalidating references
+                if (auto nh_old = s->known_connectable.extract({ info_new.listen_address(), event.port }); !std::empty(nh_old))
+                {
+                    auto& info_old = nh_old.mapped();
+                    TR_ASSERT(nh_old.key() == info_old.listen_socket_address());
+                    TR_ASSERT(nh_old.key().address() == info_new.listen_address());
+
+                    // If there is an existing connection to this peer, keep the better one
+                    if (info_old.is_connected())
+                    {
+                        if (CompareAtomsByUsefulness(info_new, info_old))
+                        {
+                            auto it = std::find_if(
+                                std::begin(s->peers),
+                                std::end(s->peers),
+                                [&info_old](tr_peerMsgs const* const peer) { return peer->peer_info == &info_old; });
+                            TR_ASSERT(it != std::end(s->peers));
+                            s->remove_peer(*it);
+                        }
+                        else
+                        {
+                            info_new.merge_connectable_into_incoming(info_old, true);
```

Comment:
The `from_first_` field of the 2 info objects need to be swapped in order for stat keeping to be correct. 💫

Reply:
f012b98